File: app/routers/mongodb.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field, validator
from typing import Dict, Any, List, Optional, Union
import logging
import json
from bson import json_util, ObjectId
import pymongo
from pymongo.errors import PyMongoError
import re
from app.utils.response import success_response, error_response
from app.utils.pool import mongodb_pool

router = APIRouter(tags=["MongoDB"])
logger = logging.getLogger("database-api")

class MongoDBConnectionInfo(BaseModel):
    host: str
    port: int = 27017
    database: str
    username: Optional[str] = None
    password: Optional[str] = None
    auth_source: Optional[str] = "admin"
    connect_timeout_ms: int = 30000

    @validator('host')
    def validate_host(cls, v):
        if not v or not v.strip():
            raise ValueError("MongoDB主机地址不能为空")
        return v

    @validator('database')
    def validate_database(cls, v):
        if not v or not v.strip():
            raise ValueError("MongoDB数据库名不能为空")
        return v

class MongoDBExecuteRequest(BaseModel):
    connection: MongoDBConnectionInfo
    collection: str
    operation: str
    filter: Optional[Dict[str, Any]] = None
    update: Optional[Dict[str, Any]] = None
    document: Optional[Dict[str, Any]] = None
    documents: Optional[List[Dict[str, Any]]] = None
    projection: Optional[Dict[str, Any]] = None
    sort: Optional[List[tuple]] = None
    limit: Optional[int] = None
    skip: Optional[int] = None
    pipeline: Optional[List[Dict[str, Any]]] = None

    @validator('collection')
    def validate_collection(cls, v):
        if not v or not v.strip():
            raise ValueError("MongoDB集合名不能为空")
        return v

    @validator('operation')
    def validate_operation(cls, v):
        if not v or not v.strip():
            raise ValueError("MongoDB操作类型不能为空")
        
        valid_operations = [
            'find', 'findone', 'insert', 'insertmany', 
            'update', 'updatemany', 'delete', 'deletemany',
            'aggregate', 'count'
        ]
        
        if v.lower() not in valid_operations:
            raise ValueError(f"不支持的操作类型: {v}。支持的操作类型有: {', '.join(valid_operations)}")
        
        return v.lower()

# 将MongoDB的BSON对象转换为JSON可序列化的格式
def parse_json(data):
    return json.loads(json_util.dumps(data))

@router.post("/mongodb")
async def execute_mongodb(request: MongoDBExecuteRequest):
    """
    执行MongoDB原生操作
    
    - 支持find, findOne, insert, insertMany, update, updateMany, delete, deleteMany等操作
    - 返回查询结果或操作影响的文档数量
    """
    client = None
    try:
        # 验证操作特定的参数
        operation = request.operation.lower()
        
        # 验证查询操作所需参数
        if operation in ["find", "findone"]:
            if request.filter is None:
                logger.warning("查询操作未提供filter参数，将使用空过滤条件")
        
        # 验证插入操作所需参数
        if operation == "insert" and not request.document:
            return error_response(1101, "插入操作缺少document字段，请提供要插入的文档")
            
        if operation == "insertmany" and not request.documents:
            return error_response(1102, "批量插入操作缺少documents字段，请提供要插入的文档列表")
        
        # 验证更新操作所需参数    
        if operation in ["update", "updatemany"]:
            if not request.filter:
                return error_response(1103, "更新操作缺少filter字段，请提供查询条件")
            if not request.update:
                return error_response(1104, "更新操作缺少update字段，请提供更新内容")
                
            # 检查更新操作是否包含操作符
            if request.update and not any(key.startswith('$') for key in request.update.keys()):
                return error_response(1105, "更新操作的update字段格式不正确，应包含至少一个更新操作符如 $set, $unset 等")
        
        # 验证删除操作所需参数
        if operation in ["delete", "deletemany"] and not request.filter:
            return error_response(1106, "删除操作缺少filter字段，请提供查询条件")
            
        # 验证聚合操作所需参数
        if operation == "aggregate" and not request.pipeline:
            return error_response(1107, "聚合操作缺少pipeline字段，请提供聚合管道")
        
        # 从连接池获取MongoDB客户端
        client, error = mongodb_pool.get_client(
            host=request.connection.host,
            port=request.connection.port,
            database=request.connection.database, 
            username=request.connection.username,
            password=request.connection.password,
            auth_source=request.connection.auth_source,
            connect_timeout_ms=request.connection.connect_timeout_ms
        )
        
        if error:
            logger.error(f"从连接池获取MongoDB客户端失败: {error}")
            error_msg = error
            # 提供更友好的连接错误消息
            if "timed out" in error_msg:
                error_msg = f"连接MongoDB服务器超时，请检查主机地址和端口是否正确: {request.connection.host}:{request.connection.port}"
            elif "not authorized" in error_msg:
                error_msg = "MongoDB认证失败，请检查用户名和密码是否正确"
            elif "Authentication failed" in error_msg:
                error_msg = "MongoDB认证失败，请检查用户名和密码是否正确"
            
            return error_response(1001, f"数据库连接错误: {error_msg}")
            
        logger.info(f"成功从连接池获取MongoDB客户端: {request.connection.host}:{request.connection.port}/{request.connection.database}")
        
        # 获取数据库和集合
        db = client[request.connection.database]
        
        try:
            # 先检查集合是否存在
            collection_exists = request.collection in db.list_collection_names()
            if not collection_exists:
                logger.warning(f"集合不存在: {request.collection}")
                # 所有操作在集合不存在时都返回错误信息
                return error_response(1120, f"集合 '{request.collection}' 不存在")
                
            collection = db[request.collection]
        except Exception as e:
            logger.error(f"获取集合失败: {str(e)}")
            return error_response(1002, f"获取集合 {request.collection} 失败，请检查集合名是否正确")
        
        # 执行操作
        try:
            # 查询操作
            if operation == "find":
                filter_dict = request.filter or {}
                projection_dict = request.projection or None
                cursor = collection.find(filter_dict, projection_dict)
                
                # 应用排序、跳过和限制
                if request.sort:
                    cursor = cursor.sort(request.sort)
                if request.skip:
                    cursor = cursor.skip(request.skip)
                if request.limit:
                    cursor = cursor.limit(request.limit)
                    
                # 获取结果并转换为JSON
                results = list(cursor)
                
                # 检查结果是否为空
                if not results:
                    logger.info(f"查询未返回任何结果: {filter_dict}")
                
                return success_response(parse_json(results))
                
            elif operation == "findone":
                filter_dict = request.filter or {}
                projection_dict = request.projection or None
                result = collection.find_one(filter_dict, projection_dict)
                
                if result:
                    return success_response([parse_json(result)])
                else:
                    logger.info(f"查询未返回任何结果: {filter_dict}")
                    return success_response([])
            
            # 插入操作
            elif operation == "insert":
                # 已在前面验证document存在
                result = collection.insert_one(request.document)
                logger.info(f"成功插入文档，ID: {result.inserted_id}")
                return success_response(1)
                
            elif operation == "insertmany":
                # 已在前面验证documents存在
                result = collection.insert_many(request.documents)
                logger.info(f"成功插入{len(result.inserted_ids)}个文档")
                return success_response(len(result.inserted_ids))
            
            # 更新操作
            elif operation == "update":
                # 已在前面验证filter和update存在
                result = collection.update_one(request.filter, request.update)
                
                if result.modified_count == 0:
                    logger.info(f"更新操作未影响任何文档: {request.filter}")
                else:
                    logger.info(f"成功更新{result.modified_count}个文档")
                
                return success_response(result.modified_count)
                
            elif operation == "updatemany":
                # 已在前面验证filter和update存在
                result = collection.update_many(request.filter, request.update)
                
                if result.modified_count == 0:
                    logger.info(f"批量更新操作未影响任何文档: {request.filter}")
                else:
                    logger.info(f"成功更新{result.modified_count}个文档")
                
                return success_response(result.modified_count)
            
            # 删除操作
            elif operation == "delete":
                # 已在前面验证filter存在
                result = collection.delete_one(request.filter)
                
                if result.deleted_count == 0:
                    logger.info(f"删除操作未影响任何文档: {request.filter}")
                else:
                    logger.info(f"成功删除{result.deleted_count}个文档")
                
                return success_response(result.deleted_count)
                
            elif operation == "deletemany":
                # 已在前面验证filter存在
                result = collection.delete_many(request.filter)
                
                if result.deleted_count == 0:
                    logger.info(f"批量删除操作未影响任何文档: {request.filter}")
                else:
                    logger.info(f"成功删除{result.deleted_count}个文档")
                
                return success_response(result.deleted_count)
                
            # 聚合操作
            elif operation == "aggregate":
                # 已在前面验证pipeline存在
                try:
                    results = list(collection.aggregate(request.pipeline))
                    
                    if not results:
                        logger.info("聚合操作未返回任何结果")
                    
                    return success_response(parse_json(results))
                except pymongo.errors.OperationFailure as e:
                    error_msg = str(e)
                    
                    if "pipeline" in error_msg:
                        error_msg = f"聚合管道格式错误: {error_msg}"
                    elif "unknown operator" in error_msg:
                        match = re.search(r'unknown operator: ([^\s]+)', error_msg)
                        if match:
                            op = match.group(1)
                            error_msg = f"聚合管道中使用了未知操作符: {op}"
                    elif "accumulator" in error_msg:
                        error_msg = f"聚合累加器错误: {error_msg}"
                    
                    logger.error(f"聚合操作失败: {error_msg}")
                    return error_response(1108, f"聚合操作失败: {error_msg}")
                
            # 计数操作
            elif operation == "count":
                filter_dict = request.filter or {}
                count = collection.count_documents(filter_dict)
                return success_response(count)
            
            else:
                # 虽然前面已验证，但保留兜底处理
                return error_response(1109, f"不支持的操作类型: {request.operation}")
                
        except pymongo.errors.DuplicateKeyError as e:
            logger.error(f"MongoDB插入重复键错误: {str(e)}")
            # 尝试提取重复键的字段名
            error_msg = str(e)
            match = re.search(r'duplicate key error .* key: \{ ([^:]+)', error_msg)
            field = match.group(1).strip() if match else "未知字段"
            return error_response(1110, f"插入操作失败，文档中的 {field} 与已有文档重复")
            
        except pymongo.errors.BulkWriteError as e:
            logger.error(f"MongoDB批量写入错误: {str(e)}")
            return error_response(1111, f"批量操作失败: {str(e)}")
            
        except pymongo.errors.WriteError as e:
            logger.error(f"MongoDB写入错误: {str(e)}")
            return error_response(1112, f"写入操作失败: {str(e)}")
            
        except pymongo.errors.InvalidOperation as e:
            logger.error(f"MongoDB无效操作: {str(e)}")
            error_msg = str(e)
            if "cannot use an empty filter" in error_msg:
                error_msg = "不能使用空的过滤条件，请提供查询条件"
            return error_response(1113, f"无效操作: {error_msg}")
            
        except pymongo.errors.DocumentTooLarge as e:
            logger.error(f"MongoDB文档过大: {str(e)}")
            return error_response(1114, "文档大小超过MongoDB限制")
            
        except Exception as e:
            logger.error(f"MongoDB操作执行错误: {str(e)}")
            return error_response(1115, f"操作执行错误: {str(e)}")
                
    except pymongo.errors.OperationFailure as e:
        logger.error(f"MongoDB操作失败: {str(e)}")
        error_msg = str(e)
        
        if "authentication failed" in error_msg.lower():
            error_msg = "MongoDB认证失败，请检查用户名和密码是否正确"
        elif "not authorized" in error_msg.lower():
            error_msg = "没有操作权限，请检查用户是否有足够的权限"
        
        return error_response(1002, f"数据库操作错误: {error_msg}")
        
    except PyMongoError as e:
        logger.error(f"MongoDB错误: {str(e)}")
        return error_response(1000, f"数据库错误: {str(e)}")
        
    except ValueError as e:
        logger.error(f"参数验证错误: {str(e)}")
        return error_response(1116, f"参数验证错误: {str(e)}")
        
    except Exception as e:
        logger.error(f"未预期的错误: {str(e)}")
        return error_response(9999, f"服务器错误: {str(e)}")
        
    finally:
        # 注意：连接池模式下不需要关闭MongoDB客户端连接
        # 客户端连接会被池管理器自动管理和重用
        logger.info("MongoDB操作完成") 

File: controller/mongodb.go
package controller

import (
	"context"
	"database-api-public-go/utils"
	"fmt"
	"log"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

// MongoDBConnectionInfo MongoDB连接信息
type MongoDBConnectionInfo struct {
	Host            string `json:"host" binding:"required"`
	Port            int    `json:"port" binding:"required"`
	Database        string `json:"database" binding:"required"`
	Username        string `json:"username"`
	Password        string `json:"password"`
	AuthSource      string `json:"auth_source"`
	ConnectTimeoutMS int   `json:"connect_timeout_ms"`
}

// MongoDBExecuteRequest MongoDB执行请求
type MongoDBExecuteRequest struct {
	Connection MongoDBConnectionInfo `json:"connection" binding:"required"`
	Collection string                `json:"collection" binding:"required"`
	Operation  string                `json:"operation" binding:"required"`
	Filter     map[string]interface{} `json:"filter"`
	Update     map[string]interface{} `json:"update"`
	Document   map[string]interface{} `json:"document"`
	Documents  []map[string]interface{} `json:"documents"`
	Projection map[string]interface{} `json:"projection"`
	Sort       []interface{}          `json:"sort"`
	Limit      int64                  `json:"limit"`
	Skip       int64                  `json:"skip"`
	Pipeline   []map[string]interface{} `json:"pipeline"`
}

// MongoDBHandler 处理MongoDB请求
func MongoDBHandler(c *gin.Context) {
	var request MongoDBExecuteRequest

	// 绑定请求数据
	if err := c.ShouldBindJSON(&request); err != nil {
		utils.BadRequestResponse(c, fmt.Sprintf("无效的请求数据: %v", err))
		return
	}

	// 验证操作类型
	operation := strings.ToLower(request.Operation)
	if !isValidOperation(operation) {
		utils.BadRequestResponse(c, "不支持的操作类型。支持的操作类型有: find, findone, insert, insertmany, update, updatemany, delete, deletemany, aggregate, count")
		return
	}

	// 验证操作参数
	if err := validateOperationParams(request); err != nil {
		utils.ErrorResponse(c, 1100, err.Error())
		return
	}

	// 设置默认值
	if request.Connection.AuthSource == "" {
		request.Connection.AuthSource = "admin"
	}
	if request.Connection.ConnectTimeoutMS == 0 {
		request.Connection.ConnectTimeoutMS = 30000
	}

	// 获取MongoDB客户端
	client, err := utils.MongoDBPool.GetClient(
		request.Connection.Host,
		request.Connection.Port,
		request.Connection.Database,
		request.Connection.Username,
		request.Connection.Password,
		request.Connection.AuthSource,
	)
	if err != nil {
		log.Printf("MongoDB连接错误: %v", err)
		utils.ErrorResponse(c, 1001, fmt.Sprintf("数据库连接错误: %v", err))
		return
	}

	// 获取数据库和集合
	db := client.Database(request.Connection.Database)
	
	// 检查集合是否存在
	collections, err := db.ListCollectionNames(context.Background(), bson.D{})
	if err != nil {
		utils.ErrorResponse(c, 1002, fmt.Sprintf("列出集合失败: %v", err))
		return
	}
	
	collectionExists := false
	for _, coll := range collections {
		if coll == request.Collection {
			collectionExists = true
			break
		}
	}
	
	if !collectionExists && operation != "insert" && operation != "insertmany" {
		utils.ErrorResponse(c, 1120, fmt.Sprintf("集合 '%s' 不存在", request.Collection))
		return
	}
	
	collection := db.Collection(request.Collection)
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	// 根据操作类型执行相应的MongoDB操作
	switch operation {
	case "find":
		handleFind(ctx, c, collection, request)
	case "findone":
		handleFindOne(ctx, c, collection, request)
	case "insert":
		handleInsert(ctx, c, collection, request)
	case "insertmany":
		handleInsertMany(ctx, c, collection, request)
	case "update":
		handleUpdate(ctx, c, collection, request, false)
	case "updatemany":
		handleUpdate(ctx, c, collection, request, true)
	case "delete":
		handleDelete(ctx, c, collection, request, false)
	case "deletemany":
		handleDelete(ctx, c, collection, request, true)
	case "aggregate":
		handleAggregate(ctx, c, collection, request)
	case "count":
		handleCount(ctx, c, collection, request)
	}
}

// 验证操作类型
func isValidOperation(operation string) bool {
	validOperations := map[string]bool{
		"find": true, "findone": true, "insert": true, "insertmany": true,
		"update": true, "updatemany": true, "delete": true, "deletemany": true,
		"aggregate": true, "count": true,
	}
	return validOperations[operation]
}

// 验证操作参数
func validateOperationParams(request MongoDBExecuteRequest) error {
	operation := strings.ToLower(request.Operation)

	// 验证插入操作参数
	if operation == "insert" && request.Document == nil {
		return fmt.Errorf("插入操作缺少document字段，请提供要插入的文档")
	}
	if operation == "insertmany" && (request.Documents == nil || len(request.Documents) == 0) {
		return fmt.Errorf("批量插入操作缺少documents字段，请提供要插入的文档列表")
	}

	// 验证更新操作参数
	if (operation == "update" || operation == "updatemany") {
		if request.Filter == nil {
			return fmt.Errorf("更新操作缺少filter字段，请提供查询条件")
		}
		if request.Update == nil {
			return fmt.Errorf("更新操作缺少update字段，请提供更新内容")
		}

		// 检查更新操作是否包含操作符
		hasUpdateOperator := false
		for key := range request.Update {
			if strings.HasPrefix(key, "$") {
				hasUpdateOperator = true
				break
			}
		}
		if !hasUpdateOperator {
			return fmt.Errorf("更新操作的update字段格式不正确，应包含至少一个更新操作符如 $set, $unset 等")
		}
	}

	// 验证删除操作参数
	if (operation == "delete" || operation == "deletemany") && request.Filter == nil {
		return fmt.Errorf("删除操作缺少filter字段，请提供查询条件")
	}

	// 验证聚合操作参数
	if operation == "aggregate" && (request.Pipeline == nil || len(request.Pipeline) == 0) {
		return fmt.Errorf("聚合操作缺少pipeline字段，请提供聚合管道")
	}

	return nil
}

// 处理查询操作
func handleFind(ctx context.Context, c *gin.Context, collection *mongo.Collection, request MongoDBExecuteRequest) {
	filter := convertMapToBSON(request.Filter)
	findOptions := options.Find()

	if request.Projection != nil {
		findOptions.SetProjection(convertMapToBSON(request.Projection))
	}
	if request.Sort != nil && len(request.Sort) > 0 {
		findOptions.SetSort(convertToSort(request.Sort))
	}
	if request.Skip > 0 {
		findOptions.SetSkip(request.Skip)
	}
	if request.Limit > 0 {
		findOptions.SetLimit(request.Limit)
	}

	cursor, err := collection.Find(ctx, filter, findOptions)
	if err != nil {
		utils.ErrorResponse(c, 1200, fmt.Sprintf("查询操作失败: %v", err))
		return
	}
	defer cursor.Close(ctx)

	var results []map[string]interface{}
	if err := cursor.All(ctx, &results); err != nil {
		utils.ErrorResponse(c, 1201, fmt.Sprintf("处理查询结果失败: %v", err))
		return
	}

	for i := range results {
		results[i] = convertBSONToMap(results[i])
	}

	utils.SuccessResponse(c, results)
}

// 处理单条查询操作
func handleFindOne(ctx context.Context, c *gin.Context, collection *mongo.Collection, request MongoDBExecuteRequest) {
	filter := convertMapToBSON(request.Filter)
	findOptions := options.FindOne()

	if request.Projection != nil {
		findOptions.SetProjection(convertMapToBSON(request.Projection))
	}

	var result map[string]interface{}
	err := collection.FindOne(ctx, filter, findOptions).Decode(&result)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			utils.SuccessResponse(c, []interface{}{})
		} else {
			utils.ErrorResponse(c, 1202, fmt.Sprintf("查询单条记录失败: %v", err))
		}
		return
	}

	utils.SuccessResponse(c, []interface{}{convertBSONToMap(result)})
}

// 处理插入操作
func handleInsert(ctx context.Context, c *gin.Context, collection *mongo.Collection, request MongoDBExecuteRequest) {
	document := convertMapToBSON(request.Document)
	
	result, err := collection.InsertOne(ctx, document)
	if err != nil {
		utils.ErrorResponse(c, 1203, fmt.Sprintf("插入文档失败: %v", err))
		return
	}

	utils.SuccessResponse(c, map[string]interface{}{
		"inserted_id": result.InsertedID,
	})
}

// 处理批量插入操作
func handleInsertMany(ctx context.Context, c *gin.Context, collection *mongo.Collection, request MongoDBExecuteRequest) {
	var documents []interface{}
	for _, doc := range request.Documents {
		documents = append(documents, convertMapToBSON(doc))
	}
	
	result, err := collection.InsertMany(ctx, documents)
	if err != nil {
		utils.ErrorResponse(c, 1204, fmt.Sprintf("批量插入文档失败: %v", err))
		return
	}

	utils.SuccessResponse(c, map[string]interface{}{
		"inserted_count": len(result.InsertedIDs),
		"inserted_ids":   result.InsertedIDs,
	})
}

// 处理更新操作
func handleUpdate(ctx context.Context, c *gin.Context, collection *mongo.Collection, request MongoDBExecuteRequest, many bool) {
	filter := convertMapToBSON(request.Filter)
	update := convertMapToBSON(request.Update)

	if many {
		result, err := collection.UpdateMany(ctx, filter, update)
		if err != nil {
			utils.ErrorResponse(c, 1205, fmt.Sprintf("批量更新文档失败: %v", err))
			return
		}
		utils.SuccessResponse(c, map[string]interface{}{
			"matched_count":  result.MatchedCount,
			"modified_count": result.ModifiedCount,
		})
	} else {
		result, err := collection.UpdateOne(ctx, filter, update)
		if err != nil {
			utils.ErrorResponse(c, 1206, fmt.Sprintf("更新文档失败: %v", err))
			return
		}
		utils.SuccessResponse(c, map[string]interface{}{
			"matched_count":  result.MatchedCount,
			"modified_count": result.ModifiedCount,
		})
	}
}

// 处理删除操作
func handleDelete(ctx context.Context, c *gin.Context, collection *mongo.Collection, request MongoDBExecuteRequest, many bool) {
	filter := convertMapToBSON(request.Filter)

	if many {
		result, err := collection.DeleteMany(ctx, filter)
		if err != nil {
			utils.ErrorResponse(c, 1207, fmt.Sprintf("批量删除文档失败: %v", err))
			return
		}
		utils.SuccessResponse(c, map[string]interface{}{
			"deleted_count": result.DeletedCount,
		})
	} else {
		result, err := collection.DeleteOne(ctx, filter)
		if err != nil {
			utils.ErrorResponse(c, 1208, fmt.Sprintf("删除文档失败: %v", err))
			return
		}
		utils.SuccessResponse(c, map[string]interface{}{
			"deleted_count": result.DeletedCount,
		})
	}
}

// 处理聚合操作
func handleAggregate(ctx context.Context, c *gin.Context, collection *mongo.Collection, request MongoDBExecuteRequest) {
	var pipeline []bson.D
	for _, stage := range request.Pipeline {
		pipeline = append(pipeline, convertMapToBSONDoc(stage))
	}

	cursor, err := collection.Aggregate(ctx, pipeline)
	if err != nil {
		utils.ErrorResponse(c, 1209, fmt.Sprintf("聚合操作失败: %v", err))
		return
	}
	defer cursor.Close(ctx)

	var results []map[string]interface{}
	if err := cursor.All(ctx, &results); err != nil {
		utils.ErrorResponse(c, 1210, fmt.Sprintf("处理聚合结果失败: %v", err))
		return
	}

	for i := range results {
		results[i] = convertBSONToMap(results[i])
	}

	utils.SuccessResponse(c, results)
}

// 处理计数操作
func handleCount(ctx context.Context, c *gin.Context, collection *mongo.Collection, request MongoDBExecuteRequest) {
	filter := convertMapToBSON(request.Filter)
	
	count, err := collection.CountDocuments(ctx, filter)
	if err != nil {
		utils.ErrorResponse(c, 1211, fmt.Sprintf("计数操作失败: %v", err))
		return
	}

	utils.SuccessResponse(c, map[string]interface{}{
		"count": count,
	})
}

// 将map转换为BSON
func convertMapToBSON(m map[string]interface{}) bson.M {
	if m == nil {
		return bson.M{}
	}
	return bson.M(m)
}

// 将map转换为BSON文档
func convertMapToBSONDoc(m map[string]interface{}) bson.D {
	var doc bson.D
	for k, v := range m {
		// 如果值是map，递归转换
		if subMap, ok := v.(map[string]interface{}); ok {
			doc = append(doc, bson.E{Key: k, Value: convertMapToBSON(subMap)})
		} else {
			doc = append(doc, bson.E{Key: k, Value: v})
		}
	}
	return doc
}

// 将BSON转换为map
func convertBSONToMap(m map[string]interface{}) map[string]interface{} {
	result := make(map[string]interface{})
	
	for k, v := range m {
		// 转换MongoDB的_id
		if k == "_id" {
			// 将ObjectID转换为纯十六进制字符串
			idStr := fmt.Sprintf("%v", v)
			// 检查是否是ObjectID形式
			if strings.HasPrefix(idStr, "ObjectID(") && strings.HasSuffix(idStr, ")") {
				// 提取引号内的十六进制字符串
				hexString := strings.TrimPrefix(strings.TrimSuffix(idStr, ")"), "ObjectID(\"")
				hexString = strings.Trim(hexString, "\"")
				result[k] = hexString
			} else {
				// 不是ObjectID格式，直接使用原值
				result[k] = idStr
			}
			continue
		}
		
		// 处理嵌套的map
		if subMap, ok := v.(map[string]interface{}); ok {
			result[k] = convertBSONToMap(subMap)
			continue
		}
		
		// 处理数组
		if arr, ok := v.([]interface{}); ok {
			convertedArr := make([]interface{}, len(arr))
			for i, item := range arr {
				if subMap, ok := item.(map[string]interface{}); ok {
					convertedArr[i] = convertBSONToMap(subMap)
				} else {
					convertedArr[i] = item
				}
			}
			result[k] = convertedArr
			continue
		}
		
		// 其他类型直接保留
		result[k] = v
	}
	
	return result
}

// 转换排序条件
func convertToSort(sort []interface{}) bson.D {
	var doc bson.D
	
	// 处理字符串形式的排序字段 ["field1", "-field2"]
	for _, field := range sort {
		if strField, ok := field.(string); ok {
			if strings.HasPrefix(strField, "-") {
				doc = append(doc, bson.E{Key: strings.TrimPrefix(strField, "-"), Value: -1})
			} else {
				doc = append(doc, bson.E{Key: strField, Value: 1})
			}
		} else if mapField, ok := field.(map[string]interface{}); ok {
			// 处理map形式的排序字段 [{"field": 1}, {"field2": -1}]
			for k, v := range mapField {
				if val, ok := v.(int); ok {
					doc = append(doc, bson.E{Key: k, Value: val})
				} else if val, ok := v.(float64); ok {
					doc = append(doc, bson.E{Key: k, Value: int(val)})
				}
			}
		}
	}
	
	return doc
} 